compiler/fir/fir2ir/src/org/jetbrains/kotlin/fir/backend/Fir2IrImplicitCastInserter.kt

```diff
@@ -209,6 +210,9 @@ class Fir2IrImplicitCastInserter(
             expectedType.isUnit -> {
                 coerceToUnitIfNeeded(type, irBuiltIns)
             }
+            valueType.isNullabilityFlexible() && valueType.containsNull() && !expectedType.acceptsNullValues() -> {
```

Comment:
For your convenience, this is the counterpart: https://github.com/JetBrains/kotlin/blob/master/compiler/ir/ir.psi2ir/src/org/jetbrains/kotlin/psi2ir/transformations/InsertImplicitCasts.kt#L352

Note that `IrType.isNullabilityFlexible()` is indeed checking if it's a flexible type and its lower/upper bound's nullability don't match.

Reply:
Yes, I confirm this is consistent with FE 1.0.